### ISSUE TITLE
Spurious merge conflicts involving small sections.

### ISSUE THREAD

**Issue by sageserpent-open** [sageserpent-open/kineticMerge#49]
This follows on from #48 . Writing the bug reproduction test for that issue revealed an undesirable merge conflict when the problem causing the assertion failure was fixed.

This merge conflict isn't related to the *that* bug, rather it is due to rather clumsy matching of a run of text that is matched on the left and right as a divergent move, but isn't matched against on the base, despite the same piece of text `THE END.` being present on all three sides. What happens instead is that a larger match `.\nTHE END.` is made across the left and right, and this is what is recorded as a divergent insertion move, as it spans two distinct files.

It is possible to set the minimum match size down to 3 tokens to make an all-sides match of just `THE END.`, but this results in the preceding `.\n` being recorded as a divergent insertion instead - what happens is that the left/right pairwise match of `.\nTHE END.` is eaten into by the all-sides match of `THE END.`, splitting off `.\n` as a one-token left/right pairwise match. This goes on to make a merge conflict in the same way as the previous example.

**Comment by sageserpent-open** [comment]
I'm not sure how best to attack this. Say the minimum number of tokens to get a match is 4 - should the unmatched three-token section then eat into the match? Or should we take the line that if we want to see the match of size 3, we should specify it? The latter seems reasonable, it does what we ask it to do that way.

The thing is, if we want matches of size 3, then eating into the larger pairwise match *shouldn't* give us a lousy one-token pairwise match as a leftover!

One more thing - need to check the divergent insertion of the `.\n`. Is it really a bad idea to flag this as a merge conflict. Suppose it wasn't treated as a pairwise match? Would that have made things any better?

**Comment by sageserpent-open** [comment]
As of d09c6afa4920a838975055e80c012de91351d804, got a spike that looks promising.

The minimum match size is set to 3 in the test to give the machinery a chance to deal with the `THE END.` section across all three sides, and that works nicely.

The conflict handling code has been tweaked so that if the base of an edit-versus-delete (or vice-versa) conflict is the source of a move (single or divergent), then it is resolved trivially to be a coincident deletion of the base and an insertion of the edit section.

Doing this avoids an unnecessary conflict (as we know the base is simply moving away, so the apparent edit is really the insertion of an unrelated section), and allows the inserted text to participate in migration.

That leads to a somewhat better merge where the renamed file looks like this:

```text

This is the obligatory zeroth linethe first line,
followed by the secondthe obligatory zeroth line.
Can you see where this is going?
Need a hint?
This was No, good - you're a quick study.
THE END.
```

The section:

```text
the first line,
followed by the second
```

is moved divergently, following the rename of the file on one branch, but moving inside the original file on the other.

Furthermore, there is a divergent insertion of the `.` resulting from the larger pairwise match `.\nTHE END.` being eaten into.

Both the divergent move and insertion cause the original file to *pick up* content on one branch, whereas it is completely deleted on the other, so this results in a conflict with the original file's new content being left in the working tree of the merge.

As for the renamed file, the section `the obligatory zeroth line` is anchored to the preceding `This is` and to the succeeding `.\nCan you see where  ... ?\nNeed a hint?`. The latter is taken to have moved past `the first line,\nfollowed by the second` in the renamed file, so the insertions follows the non-parallel movement of both anchors into the renamed file and ends up being duplicated.

**Comment by sageserpent-open** [comment]
For context, here are the files...

BASE: (aFile.txt)

```text

This is the first line,
followed by the second.

Can you see where this is going?
Need a hint?
THE END.
```

LEFT (aFile.txt)

```text

This is the obligatory zeroth line.
Can you see where this is going?
Need a hint?
This was the first line,
followed by the second.

THE END.
```

RIGHT (theRenamedFile.txt)

```text

This is the first line,
followed by the second.

Can you see where this is going?
Need a hint? No, good - you're a quick study.
THE END.
```

**Comment by sageserpent-open** [comment]
Note that the section `This was ` was treated as an insertion and migrated using its *preceding anchor* of `.\nCan you see where  ... ?\nNeed a hint?`. It would have been nicer from a grammatical point of view if it had anchored to the succeeding half of the divergent insertion of `the first line,\nfollowed by the second`, but as a general principle, insertions (at least one-sided ones) should not be used as anchors - we need *sources* rather than destinations of moves.

As it turned out, the `This is ` section that we know as human beings was edited into `This was ` turned out to be matched anyway by another `This is ` that formed part of the *text* `This is the obligatory zeroth line.`, so Kinetic Merge happily tooks that to be a straight match. Seems fair enough.

**Comment by sageserpent-open** [comment]
Note that the divergent move of `the first line,\nfollowed by the second` is an example of #32, so we can revisit this in that ticket.

Not sure about the divergent insertion of `.\n`. Perhaps it could be handled as a degenerate case of #32?

**Comment by sageserpent-open** [comment]
For now, I'm going to leave `MainTest.issue48BugReproduction` as of d09c6afa4920a838975055e80c012de91351d804 and add additional lower-level testing to `MatchesContextTest`...

**Comment by sageserpent-open** [comment]
Tests updated, merged in 21c8d2349a027f8f9d873f4fa23a67d929ed6529.
